app/services/mail_rate_limiter.rb

```diff
@@ -0,0 +1,32 @@
+class MailRateLimiter
+  attr_reader :delay, :current_window
+
+  def send_with_delay(mail)
+    if current_window_full?
+      @delay += @window
+    end
+    if current_window_full? || current_window_expired?
+      @current_window = { started_at: Time.zone.now.utc, sent: 0 }
+    end
+    @current_window[:sent] += 1
+
+    mail.deliver_later(wait: delay)
+  end
+
+  private
+
+  def initialize(limit:, window:)
+    @limit = limit
+    @window = window
+    @current_window = { started_at: Time.zone.now.utc, sent: 0 }
+    @delay = 0
+  end
+
+  def current_window_expired?
+    @current_window[:started_at] + @window <= Time.zone.now.utc
```

Comment:
peut-être question de goût et d'habitude mais j'ai l'impression de moins faire de calcul / moins me tromper comme ceci : 
```suggestion
    (@current_window[:started_at] + @window).past?
```